TombEngine/Objects/TR3/Trap/ElectricCleaner.cpp

```diff
@@ -0,0 +1,426 @@
+					item.ItemFlags[1] &= ~(1 << 1); // Turn off 1st bit for flagTurnRight. (So it'll turn to the left)
+				
+				//Store goal angle to control the rotation.
+				item.ItemFlags[6] = FROM_RAD(atan2(NewDirection.x, NewDirection.z));
+
+				if (item.Pose.Orientation.y - item.ItemFlags[6] == 0)
+					//If it doesn't have to rotate, do forward movement to keep smooth movement.
+					item.Pose.Position = item.Pose.Position + ForwardDirection * moveVel;
+				else
+					//If it has to rotate, stop detection so it doesn't calculate collisions again while rotating in the same sector.
+					item.ItemFlags[1] &= ~(1 << 0); // Turn off 1st bit for flagDoDetection.
+			}
+			else
+			{
+				item.Pose.Position.y = col.Position.Floor;
+
+				//Is not in the center of a tile, keep moving forward. 
+				item.Pose.Position = item.Pose.Position + ForwardDirection * moveVel;
+
+				auto slope = col.Block->FloorSlope(0);
+
+				if (slope.LengthSquared() > 0) //If it's a slope, don't do turns.
+					item.ItemFlags[1] &= ~(1 << 0);	// Turn off 1st bit for flagDoDetection.
+				else
+					item.ItemFlags[1] |= (1 << 0);	// Turn on 1st bit for flagDoDetection.
+			}
+		}
+
+		AnimateItem(&item);
+
+		int probedRoomNumber = GetCollision(&item).RoomNumber;
+		if (item.RoomNumber != probedRoomNumber)
+			ItemNewRoom(itemNumber, probedRoomNumber);
+
+		auto radius = Vector2(object.radius, object.radius);
+		AlignEntityToSurface(&item, radius);
+
+		SpawnElectricCleanerSparks(item);
+		ElectricCleanerToItemCollision(item);
+	}
+
+	bool IsNextSectorValid(ItemInfo& item, const Vector3& Dir)
+	{
+		GameVector detectionPoint = item.Pose.Position + Dir * BLOCK(1);
+		detectionPoint.RoomNumber = item.RoomNumber;
+
+		auto col = GetCollision(detectionPoint);
+
+		//Is a wall
+		if (col.Block->IsWall(detectionPoint.x, detectionPoint.z))
+			return false;
+
+		//Is it a sliding slope?
+		if (col.Position.FloorSlope)
+			return false;
+
+		if (abs(col.FloorTilt.x) == 0 && abs(col.FloorTilt.y) == 0) //Is a flat tile
+		{
+			//Is a 1 click step (higher or lower).
+			int distanceToFloor = abs(col.Position.Floor - item.Pose.Position.y);
+			if (distanceToFloor >= CLICK(1))
+				return false;
+		}
+		else //Is a slope tile
+		{
+			//Is a 2 click step (higher or lower).
+			int distanceToFloor = abs(col.Position.Floor - item.Pose.Position.y);
+			if (distanceToFloor > CLICK(2))
+				return false;
+
+			short slopeAngle = ANGLE(0.0f);
+
+			if (col.FloorTilt.x > 0)
+				slopeAngle = -ANGLE(90.0f);
+			else if (col.FloorTilt.x < 0)
+				slopeAngle = ANGLE(90.0f);
+
+			if (col.FloorTilt.y > 0 && col.FloorTilt.y > abs(col.FloorTilt.x))
+				slopeAngle = ANGLE(180.0f);
+			else if (col.FloorTilt.y < 0 && -col.FloorTilt.y > abs(col.FloorTilt.x))
+				slopeAngle = ANGLE(0.0f);
+
+			auto angleDir = FROM_RAD(atan2(Dir.x, Dir.z));
+			auto alignment = slopeAngle - angleDir;
+
+			//Is slope not aligned with the direction?
+			if ((alignment != 32768) && (alignment != 0) && (alignment != -32768))
```

Comment:
Is it possible to demagic these angle values or are they like to avoid bugs?

Reply:
I am not sure about the conversion. I know that:
- 32768 (in decimal) 
- is the same that 8000 in hexadecimal
- and the same that Lara's angleº + 180º

0 is Lara's angle, and -32768 I had to add it so it can consider also the slopes in opossite direction.